File: app/src/main/java/tanzent/cassette/request/PlayListUriRequest.java
package tanzent.cassette.request;

import android.net.Uri;

import com.facebook.drawee.view.SimpleDraweeView;

import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;
import io.reactivex.observers.DisposableObserver;
import tanzent.cassette.request.network.RxUtil;
import tanzent.cassette.util.PlayListUtil;

import static tanzent.cassette.util.ImageUriUtil.getSearchRequestWithAlbumType;

/**
 * Created by Remix on 2017/11/30.
 */

public class PlayListUriRequest extends LibraryUriRequest {
    public PlayListUriRequest(SimpleDraweeView image, UriRequest request, RequestConfig config) {
        super(image, request, config);
    }

    @Override
    public void onError(String errMsg) {
//        mImage.setImageURI(Uri.EMPTY);
    }

    @Override
    public Disposable load() {
        return Observable.concat(
                getCustomThumbObservable(mRequest),
                Observable.fromIterable(PlayListUtil.getMP3ListByIds(PlayListUtil.getSongIds(mRequest.getID()), mRequest.getID()))
                        .concatMapDelayError(song -> getCoverObservable(getSearchRequestWithAlbumType(song))))
                .firstOrError()
                .toObservable()
                .compose(RxUtil.applyScheduler())
                .subscribeWith(new DisposableObserver<String>() {
                    @Override
                    protected void onStart() {
                        mImage.setImageURI(Uri.EMPTY);
                    }

                    @Override
                    public void onNext(String s) {
                        onSuccess(s);
                    }

                    @Override
                    public void onError(Throwable e) {
                        PlayListUriRequest.this.onError(e.toString());
                    }

                    @Override
                    public void onComplete() {

                    }
                });
    }

}

File: app/src/main/java/tanzent/cassette/request/PlayListUriRequest.kt
package tanzent.cassette.request

import android.net.Uri
import com.facebook.drawee.view.SimpleDraweeView
import io.reactivex.Observable
import io.reactivex.ObservableSource
import io.reactivex.SingleSource
import io.reactivex.disposables.Disposable
import io.reactivex.functions.Function
import io.reactivex.observers.DisposableObserver
import tanzent.cassette.App
import tanzent.cassette.bean.mp3.Song
import tanzent.cassette.db.room.DatabaseRepository
import tanzent.cassette.db.room.model.PlayList
import tanzent.cassette.request.network.RxUtil
import tanzent.cassette.util.ImageUriUtil.getSearchRequestWithAlbumType

/**
 * Created by Remix on 2017/11/30.
 */

open class PlayListUriRequest(image: SimpleDraweeView, request: UriRequest, config: RequestConfig) : LibraryUriRequest(image, request, config) {

  override fun onError(throwable: Throwable?) {
    super.onError(throwable)
//    mImageRef.get()?.setImageURI(Uri.EMPTY)
  }

  override fun load(): Disposable {
    val coverObservables = DatabaseRepository.getInstance()
        .getPlayList(mRequest.id)
        .flatMap(Function<PlayList, SingleSource<List<Song>>> { playList ->
          DatabaseRepository.getInstance()
              .getPlayListSongs(App.getContext(), playList, true)
        })
        .flatMapObservable(Function<List<Song>, ObservableSource<Song>> { songs ->
          Observable.create { emitter ->
            for (song in songs) {
              emitter.onNext(song)
            }
            emitter.onComplete()
          }
        })
        .concatMapDelayError(Function<Song, ObservableSource<String>> { song ->
          getCoverObservable(getSearchRequestWithAlbumType(song)) })

    return Observable.concat(getCustomThumbObservable(mRequest), coverObservables)
        .firstOrError()
        .toObservable()
        .compose(RxUtil.applyScheduler())
        .subscribeWith(object : DisposableObserver<String>() {
          override fun onStart() {
            mImageRef.get()?.setImageURI(Uri.EMPTY)
          }

          override fun onNext(s: String) {
            onSuccess(s)
          }

          override fun onError(e: Throwable) {
            this@PlayListUriRequest.onError(e)
          }

          override fun onComplete() {

          }
        })
  }

}
